backend/cn/lib/builtins.ml

```diff
@@ -61,11 +85,48 @@ let max_bits_def (sign, n) =
     | Signed -> (Z.(shift_left one (Int.sub n 1) - one), "i")
   in
   let name = "MAX" ^ letter ^ Int.to_string n in
-  ( name,
-    Sym.fresh_named name,
-    mk_arg0 (fun loc -> IT.sterm_of_term @@ num_lit_ num (BT.Bits (sign, n)) loc) )
+  num_lit_ num (BT.Bits (sign, n)) builtin_location |> mk_builtin_arg0 name
+
+
+let max_min_bits =
+  let sizes = [ 8; 16; 32; 64 ] in
+  let signs = [ BT.Unsigned; Signed ] in
+  List.fold_left
+    (fun acc sign ->
+      List.fold_left
+        (fun acc size -> max_bits_def (sign, size) :: min_bits_def (sign, size) :: acc)
+        acc
+        sizes)
+    []
+    signs
+
 
+let not_def =
+  mk_builtin_arg1 "not" BT.Bool (fun (sym, t) ->
+    not_ (sym_ (sym, t, builtin_location)) builtin_location)
 
+
+let is_null_def : string * Sym.t * builtin_fn =
+  mk_builtin_arg1 "is_null" BT.Loc (fun (sym, t) ->
+    (sym_ (sym, t, builtin_location) %== null_ builtin_location) builtin_location)
+
+
+let ptr_eq_def : string * Sym.t * builtin_fn =
+  var_binop EQ BT.Bool |> mk_builtin_arg2 "ptr_eq" (BT.Loc, BT.Loc)
+
+
+let prov_eq_def : string * Sym.t * builtin_fn =
+  var_binop EQ BT.Bool |> mk_builtin_arg2 "prov_eq" (BT.Loc, BT.Loc)
```

Comment:
The types are correct but the implementation needs to cast the arguments to `allocId_` before calling equal.